lnst/Recipes/ENRT/ConfigMixins/FirewallMixin.py

```diff
@@ -0,0 +1,197 @@
+from lnst.Recipes.ENRT.ConfigMixins import BaseSubConfigMixin
+from tempfile import mkstemp
+from os import close, write, unlink
+import logging
+import copy
+
+class FirewallMixin(BaseSubConfigMixin):
+    """
+    A config mixin to apply custom firewall rulesets on hosts.
+    Do not inherit directly, use one of the derived classes below instead.
+    """
+
+    @property
+    def firewall_rulesets(self):
+        """
+        This property holds a dictionary of firewall rulesets in textual
+        representation, indexed by the host it should be applied to. A typical
+        use is:
+
+        { self.matched.host1: <host1 ruleset>,
+          self.matched.host2: <host2 ruleset> }
+
+        The rulesets will be applied by a derived class's _apply_ruleset()
+        method, i.e. typically fed into 'nft -f' or 'iptables-restore'.
+        """
+        return {}
+
+    def _apply_ruleset(self, host, path):
+        raise NotImplementedError("Method _apply_ruleset must be defined by a child class.")
+
+    def _dump_ruleset(self, host, config):
+        raise NotImplementedError("Method _dump_ruleset must be defined by a child class.")
+
+    def _push_ruleset(self, host, ruleset):
+        """
+        Push `ruleset` into a temporary file on `host`, returning its path.
+        """
+        fd, name = mkstemp(prefix="FirewallMixin", text=True)
+        write(fd, ruleset.encode('utf-8'))
+        close(fd)
+        remote_path = host.copy_file_to_machine(name)
+        unlink(name)
+        return remote_path
+
+    def apply_sub_configuration(self, config):
+        super().apply_sub_configuration(config)
+
+        if not config.firewall_rulesets:
+            config.firewall_rulesets = self.firewall_rulesets
+
+        self.stored_rulesets = {}
+        for host, ruleset in config.firewall_rulesets.items():
+            self.stored_rulesets[host] = self._dump_ruleset(host, config)
+
+            name = self._push_ruleset(host, ruleset)
+            self._apply_ruleset(host, name)
+            host.run('rm -f "%s"' % name)
```

Comment:
Maybe use more convenient way of f-strings:
```python
host.run(f"rm -f \"{name}\"")
```

The expression in curly brackets can be also a function call.

Reply:
this is what removes the pushed files - i think this answers your other comment in https://github.com/LNST-project/lnst/pull/364/files#r1519638084

Reply:
I converted the code to using f-strings everywhere. My Python-Fu is ~15 years old.